samples/generic-synthetic-broken-links/brokenLinkChecker.js

```diff
@@ -0,0 +1,136 @@
+    const stringified_dict = Object.entries(non2xxLinks).map(([key, value]) => `${key}:${value}`).join(", ");
+    throw new Error("Non 2xx Links: " + stringified_dict);
+  }
+}
+
+/**
+ * Retrieves all links on the page using Puppeteer, removing duplicate links and
+ * those ending with '#', to ensure reliable navigation within Puppeteer. Only follow links that start with http or https
+ *
+ * @param {Page} page - The Puppeteer page instance to retrieve the links from.
+ * @returns {Array<string>} - An array of the target URLs as strings.
+ */
+async function retrieveLinks(page) {
+  const links = await page.evaluate(() => {
+    const anchors = Array.from(document.querySelectorAll('a'));
+    return anchors.map(anchor => anchor.href);
+  });
+
+  // Filter out duplicate links and remove trailing "#"
+  const filteredLinks = [...new Set(
+    links.map(link => link.toString().replace(/#$/, ''))
+  )].filter(link => link.startsWith('http') || link.startsWith('https'));
+
+  return filteredLinks;
+}
+
+
+/**
+ * Checks the status of `maxNumberOfFollowedLinks` links scraped on `startUrl`
+ * webpage leveraging Puppeteer.
+ *
+ * @param {string} startUrl - The starting URL to check the links on.
+ * @param {number} maxNumberOfFollowedLinks - The number of links to check.
+ * @param {number} maxTimeout - The maximum timeout for each link request in ms.
+ * @returns {Promise<Array<Object>>} - Array of link status results.
+ */
+async function checkLinks(startUrl, maxNumberOfFollowedLinks, maxTimeout) {
+  const browser = await puppeteer.launch({
+    headless: "new"
+  })
+  let page = await browser.newPage();
+  const linksToReturn = [];
+  const linksToCheck = [];
+
+  // load initial page, return any error in LinkResult format
+  try {
+    // Navigate to the URL and wait for the page to finish loading
+    const parentResponse = await page.goto(
+      startUrl,
+      { waitUntil: 'load', timeout: maxTimeout });
+    linksToReturn.push({
+      targetUrl: startUrl,
+      statusCode: parentResponse.status(),
+    });
+
+  } catch (error) {
+    const error_message =
+      'Could not successfully navigate to startUrl. Error: ' + error.message;
+    throw new Error(error_message);
+  };
+
+  // retrive all links from page
+  try {
+    linksToCheck.push(...await retrieveLinks(page));
+  } catch (error) {
+    const error_message =
+      'Could not successfully retrieve links. Error: ' + error.message;
+    throw new Error(error_message);
+  }
+
+  linksToCheck.splice(maxNumberOfFollowedLinks);
+
+  page = await browser.newPage();
+  page.setCacheEnabled(false); // prevents 304 errors
+  page.setDefaultNavigationTimeout(maxTimeout);
+
+  for (const targetLink of linksToCheck) {
+    try {
+      const response = await page.goto(
+        targetLink,
+        { waitUtil: 'load', timeout: maxTimeout }
+      );
+
+      linksToReturn.push({
+        targetUrl: targetLink,
+        statusCode: response.status(),
+      });
+    } catch (error) {
```

Comment:
Same comment as above re: what errors one may encounter here.

Reply:
Same as above:
1. Improperly formatted start_url:, e.g.  `Protocol error (Page.navigate): Cannot navigate to invalid URL`
2. Timeouts e.g.: `Error: Navigation timeout of 1 ms exceeded`
3. Errors thrown by `start_page.goto` call (https://pptr.dev/api/puppeteer.page.goto#remarks)

Reply:
I think that in this case we would want to have some information regarding an error scenario. Can we consider adding under linksToReturn to include the error message? Otherwise the url will fail in this weird, silent way.

Reply:
done. appropriately updated in `checkForNon2xxLinks()`